README.rst
===
hex
===

.. image:: https://travis-ci.org/steinwurf/hex.svg?branch=master
    :target: https://travis-ci.org/steinwurf/hex
    
hex is a small header only library which provides a class called dump.
dump is a class that wraps a pointer and a size and provides an output stream
operator to print the content as a hexdump.

Use as Dependency in CMake
==========================

To depend on this project when using the CMake build system, add the following
in your CMake build script:

::

   add_subdirectory("/path/to/hex" hex)
   target_link_libraries(<my_target> steinwurf::hex)

Where ``<my_target>`` is replaced by your target.

License
=======

hex is available under the BSD license.


NEWS.rst
News for hex
============

This file lists the major changes between versions. For a more detailed list of
every change, see the Git log.

Latest
------
* tbd

4.6.0
-----
* Minor: Minor changes to CMakeLists.txt.

4.5.0
-----
* Minor: Added install step to CMake.

4.4.0
-----
* Minor: Lower CMake version requirement.

4.3.0
-----
* Minor: Improve CMake build file.

4.2.0
-----
* Minor: Added CMake support.

4.1.1
-----
* Patch: Fix version name.

4.1.0
-----
* Minor: Added a constructor to create a hex::dump object from std::vector.
* Minor: Added the parse_hex_string helper function to create a byte vector
  from a hexadecimal string representation.

4.0.0
-----
* Major: Upgrade to waf-tools 4
* Minor: Upgrade to gtest 4

3.0.0
-----
* Major: Correct use flag from ``hex`` to ``hex_includes``.
* Patch: Fix naming of test files.

2.0.0
-----
* Major: Removed dependency on storage to ease reuseability.

1.0.0
-----
* Major: Initial release.


test/hexdump_test.cpp
// Copyright (c) 2016 Steinwurf ApS
// All Rights Reserved
//
// Distributed under the "BSD License". See the accompanying LICENSE.rst file.

#include <cstdint>
#include <ctime>

#include <gtest/gtest.h>

GTEST_API_ int main(int argc, char** argv)
{
    srand(static_cast<uint32_t>(time(0)));

    testing::InitGoogleTest(&argc, argv);

    return RUN_ALL_TESTS();
}


test/src/test_dump.cpp
// Copyright (c) 2016 Steinwurf ApS
// All Rights Reserved
//
// Distributed under the "BSD License". See the accompanying LICENSE.rst file.

#include <hex/dump.hpp>

#include <sstream>
#include <gtest/gtest.h>


TEST(test_dump, api)
{
    {
        std::string test = "hellohellohelloo";

        hex::dump hex((uint8_t*)test.c_str(), test.size());
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  68 65 6c 6c 6f 68 65 6c 6c 6f 68 65 6c 6c 6f 6f  "
                  "hellohellohelloo\n", ss.str());
    }

    {
        std::vector<uint8_t> data1 =
            {
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
            };

        hex::dump hex(data1.data(), data1.size());
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05 06 07 08 01 02 03 04 05 06 07 08  "
                  "................\n", ss.str());

        auto data2 = hex::parse_hex_string(
            "01 02 03 04 05 06 07 08 01 02 03 04 05 06 07 08");

        EXPECT_EQ(data1, data2);
        EXPECT_EQ(hex, hex::dump(data2));
    }

    {
        std::vector<uint8_t> data1 =
            {
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0xff, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x05, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0xab, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
            };

        hex::dump hex(data1);
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05 06 07 08 ff 02 03 04 05 06 07 08  "
                  "................\n"
                  "0010  05 02 03 04 05 06 07 08 ab 02 03 04 05 06 07 08  "
                  "................\n",
                  ss.str());

        auto data2 = hex::parse_hex_string(
            "01 02 03 04 05 06 07 08 ff 02 03 04 05 06 07 08 "
            "05 02 03 04 05 06 07 08 ab 02 03 04 05 06 07 08");

        EXPECT_EQ(data1, data2);
        EXPECT_EQ(hex, hex::dump(data2));
    }
}

TEST(test_dump, set_prefix_postfix)
{
    {
        std::string test = "hellohellohelloo";

        hex::dump hex((uint8_t*)test.c_str(), test.size());
        hex.set_prefix("0x");
        hex.set_postfix(", ");
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x68, 0x65, 0x6c, 0x6c, "
                  "0x6f, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x6f,  "
                  "hellohellohelloo\n", ss.str());
    }
    {
        std::string test = "hellohello";

        hex::dump hex((uint8_t*)test.c_str(), test.size());
        hex.set_prefix("0x");
        hex.set_postfix(", ");
        std::stringstream ss;

        ss << hex;
        auto expected =
            "0000  0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x68, 0x65, 0x6c, 0x6c, 0x6f, "
            "                                     hellohello\n";
        EXPECT_EQ(expected, ss.str());
    }
    {
        std::vector<uint8_t> data =
            {
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            };
        hex::dump hex(data);
        hex.set_prefix("0x");
        hex.set_postfix(", ");
        std::stringstream ss;

        ss << hex;

        auto expected =
            "0000  "
            "0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, "
            "0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, "
            " ................\n"
            "0010  "
            "0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, "
            "                                                 ........\n";

        EXPECT_EQ(expected, ss.str());
    }
}

TEST(test_dump, set_size)
{
    {
        // 48 bytes
        std::vector<uint8_t> data =
            {
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x05, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
                0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
            };

        hex::dump hex(data.data(), data.size());
        hex.set_size(5);

        std::stringstream ss;
        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05                                   "
                  ".....\n"
                  "....\n"
                  "0030\n",
                  ss.str());
    }

    {
        // 48 bytes
        std::vector<uint8_t> data =
            {
                0x01, 0x02, 0x03, 0x04, 0x05
            };

        hex::dump hex(data.data(), data.size());
        hex.set_size(3);

        std::stringstream ss;
        ss << hex;

        EXPECT_EQ("0000  01 02 03                                         "
                  "...\n",
                  ss.str());
    }
}


test/src/test_hexdump.cpp
// Copyright (c) 2016 Steinwurf ApS
// All Rights Reserved
//
// Distributed under the "BSD License". See the accompanying LICENSE.rst file.

#include <hex/dump.hpp>

#include <sstream>
#include <gtest/gtest.h>


TEST(test_hexdump, api)
{
    {
        std::string test = "hellohellohelloo";

        hex::dump hex((uint8_t*)test.c_str(), test.size());

        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  68 65 6c 6c 6f 68 65 6c 6c 6f 68 65 6c 6c 6f 6f  "
                  "hellohellohelloo\n", ss.str());
    }

    {
        std::vector<uint8_t> data =
        {
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
        };

        hex::dump hex(data.data(), data.size());
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05 06 07 08 01 02 03 04 05 06 07 08  "
                  "................\n", ss.str());
    }

    {
        std::vector<uint8_t> data =
        {
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x05, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
        };

        hex::dump hex(data.data(), data.size());
        std::stringstream ss;

        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05 06 07 08 01 02 03 04 05 06 07 08  "
                  "................\n"
                  "0010  05 02 03 04 05 06 07 08 01 02 03 04 05 06 07 08  "
                  "................\n",
                  ss.str());
    }
}

TEST(test_hexdump, set_size)
{
    {
        // 48 bytes
        std::vector<uint8_t> data =
        {
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x05, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08
        };

        hex::dump hex(data.data(), data.size());
        hex.set_size(5);

        std::stringstream ss;
        ss << hex;

        EXPECT_EQ("0000  01 02 03 04 05                                   "
                  ".....\n"
                  "....\n"
                  "0030\n",
                  ss.str());
    }

    {
        // 48 bytes
        std::vector<uint8_t> data =
        {
            0x01, 0x02, 0x03, 0x04, 0x05
        };

        hex::dump hex(data.data(), data.size());
        hex.set_size(3);

        std::stringstream ss;
        ss << hex;

        EXPECT_EQ("0000  01 02 03                                         "
                  "...\n",
                  ss.str());
    }
}


config.py
#!/usr/bin/env python
# encoding: utf-8

import traceback
import sys

try:
    input = raw_input
except NameError:
    pass

project_name = 'hex'
project_dependencies = \
    [
        'waf-tools',
        'gtest',
    ]


def importCode(code, name, add_to_sys_modules=0):
    """
    Import dynamically generated code as a module.
    Python recipe from http://code.activestate.com/recipes/82234
    """
    import imp

    module = imp.new_module(name)

    exec(code, module.__dict__)
    if add_to_sys_modules:
        sys.modules[name] = module

    return module


if __name__ == '__main__':
    print('Updating Smart Project Config Tool...')

    url = "https://raw.github.com/steinwurf/steinwurf-labs/" \
          "master/config_helper/config-impl.py"

    try:
        from urllib.request import urlopen, Request
    except ImportError:
        from urllib2 import urlopen, Request

    try:
        # Fetch the code file from the given url
        req = Request(url)
        response = urlopen(req)
        code = response.read()
        print("Update complete. Code size: {}\n".format(len(code)))
        try:
            # Import the code string as a module
            mod = importCode(code, "config_helper")
            # Run the actual config tool from the dynamic module
            mod.config_tool(project_dependencies, project_name)
        except:
            print("Unexpected error:")
            print(traceback.format_exc())
    except Exception as e:
        print("Could not fetch code file from:\n\t{}".format(url))
        print(e)

    input('Press ENTER to exit...')


src/hexdump/hexdump.hpp
// Copyright (c) 2016 Steinwurf ApS
// All Rights Reserved
//
// Distributed under the "BSD License". See the accompanying LICENSE.rst file.

#pragma once

#include <cassert>
#include <cstdint>
#include <ostream>
#include <iomanip>

#include <sak/storage.hpp>


namespace hexdump
{
    /// @ingroup trace
    ///
    /// @brief Small class which wraps a storage object and
    ///        provides an output stream operator to print the content
    ///        of that object as a hexdump.
    ///
    /// The hexdump format consist of three columns per row of
    /// bytes. The first column is the byte offset of the first byte
    /// in the row, the second column is the byte values in hex of
    /// those bytes (16 bytes per row) and the third column is the
    /// ascii representation of those bytes.
    ///
    /// Example of two rows (we removed the middle 6 bytes due to space
    /// limitations):
    ///
    /// @code
    /// 0000  be 40 04 71 45 XXX cd 90 e5 51 31  .@.qE XXX ...Q1
    /// 0010  9d 41 4f 37 05 XXX a9 d5 1e c7 93  .AO7. XXX .....
    /// @endcode
    struct hexdump
    {
        /// @param storage The storage object which we want to dump
        hexdump(const sak::const_storage& storage) :
            m_storage(storage),
            m_max_size(storage.m_size)
        {
            assert(m_storage.m_data);
            assert(m_storage.m_size);
            assert(m_max_size);
        }

        /// @param max_size The maximum size in bytes we wish to
        /// dump. This value can be larger than the storage object
        /// size but not zero. If max_size is smaller than the storage
        /// object size then only the max_size bytes will be printed.
        void set_max_size(uint32_t max_size)
        {
            assert(max_size > 0);
            m_max_size = max_size;
        }

        /// The storage that should be printed
        sak::const_storage m_storage;

        /// The maximum number of bytes that should be printed
        uint32_t m_max_size;
    };

    /// @ingroup trace
    ///
    /// The actual output operator which prints the storage buffer to
    /// the choosen output stream.
    ///
    /// @param out The output stream
    ///
    /// @param hex The hexdump struct initialized with the storage
    ///        that should be printed
    ///
    /// @return the used output stream
    inline std::ostream& operator<<(std::ostream& out, const hexdump& hex)
    {
        const uint8_t* data = hex.m_storage.m_data;
        uint32_t size = std::min(hex.m_storage.m_size, hex.m_max_size);

        assert(data);
        assert(size > 0);

        // don't change formatting for out
        std::ostream s(out.rdbuf());
        s << std::hex << std::setfill('0');

        std::string buf;
        buf.reserve(17); // premature optimization

        for (uint32_t i = 0; i < size; ++i)
        {
            if ((i % 16) == 0)
            {
                if (i)
                {
                    s << "  " << buf << std::endl;
                    buf.clear();
                }
                s << std::setw(4) << i << ' ';
            }

            uint8_t c = data[i];

            s << ' ' << std::setw(2) << (uint32_t) c;
            buf += (0x20 <= c && c <= 0x7e) ? c : '.';
        }

        if (size % 16)
        {
            // If size if not a multiple of 16 there will be some empty
            // columns in our print out. The remainder i.e. 16 - (size %
            // 16) e.g. for some basic cases:
            //
            // size = 15
            // remainder = 16 - (15 % 16) = 1
            //
            // size = 17
            // remainder = 16 - (17 % 16) = 15
            //
            // and so forth.
            uint32_t remainder = 16 - (size % 16);

            // We add 3 space for each missing character in the output
            s << std::string(3 * remainder, ' ');
        }

        s << "  " << buf << std::endl;

        if (size < hex.m_storage.m_size && 16 < hex.m_storage.m_size)
        {
            uint32_t last_row = (hex.m_storage.m_size / 16) * 16;

            s << "...." << std::endl;
            s << std::setw(4) << last_row << std::endl;
        }

        return out;
    }
}


src/hex/dump.hpp
// Copyright (c) 2016 Steinwurf ApS
// All Rights Reserved
//
// Distributed under the "BSD License". See the accompanying LICENSE.rst file.

#pragma once

#include <cassert>
#include <cstdint>
#include <algorithm>
#include <ostream>
#include <sstream>
#include <iomanip>
#include <string>
#include <vector>

namespace hex
{
/// @brief Small class which wraps a pointer and a size and
///        provides an output stream operator to print the content
///        as a hexdump.
///
/// The hexdump format consist of three columns per row of
/// bytes. The first column is the byte offset of the first byte
/// in the row, the second column is the byte values in hex of
/// those bytes (16 bytes per row) and the third column is the
/// ascii representation of those bytes.
///
/// Example of two rows (we removed the middle 6 bytes due to space
/// limitations):
///
/// @code
/// 0000  be 40 04 71 45 XXX cd 90 e5 51 31  .@.qE XXX ...Q1
/// 0010  9d 41 4f 37 05 XXX a9 d5 1e c7 93  .AO7. XXX .....
/// @endcode
class dump
{
    // befriend operators
    friend std::ostream& operator<<(std::ostream& out, const dump& hex);
    friend bool operator==(const dump& a, const dump& b);

public:
    /// @param data The pointer to the data which we want to dump
    /// @param max_size The maximum size in bytes of the data which we want
    ///        to dump.
    dump(const uint8_t* data, uint32_t max_size) :
        m_data(data),
        m_max_size(max_size),
        m_size(max_size)
    {
        assert(m_data);
        assert(m_max_size);
        assert(m_size);
    }

    /// @param vector The vector instance which we want to dump
    template<class PodType, class Allocator>
    dump(const std::vector<PodType, Allocator>& vector) :
        dump(vector.data(), (uint32_t) vector.size()*sizeof(PodType))
    { }

    /// @param prefix The prefix to insert before each element.
    void set_prefix(const std::string& prefix)
    {
        m_prefix = prefix;
    }

    /// @param postfix The prefix to insert after each element.
    void set_postfix(const std::string& postfix)
    {
        m_postfix = postfix;
    }

    /// @param size The size in bytes we wish to dump.
    /// The actual number of bytes dumped will equal the smallest value of
    /// either size or the max_size which was given as a parameter to
    /// the this object's constructor.
    void set_size(uint32_t size)
    {
        assert(m_size > 0);
        m_size = size;
    }

private:

    /// The pointer to the data that should be printed
    const uint8_t* m_data;

    /// The maximum number of bytes that can be printed
    const uint32_t m_max_size;

    /// The number of bytes that the user wants to print
    uint32_t m_size;

    /// The prefix to be inserted before each element
    std::string m_prefix = "";

    /// The postfix to be inserted after each element
    std::string m_postfix = " ";
};

/// Helper function to convert a hexadecimal string to an equivalent byte
/// vector. For example, the string "ff 02 03 04" is converted to a
/// std::vector<uint8_t> containing { 0xff, 0x02, 0x03, 0x04 }.
///
/// @param hex_string The input string of hexadecimal characters
/// @return The byte vector created from the string representation
inline std::vector<uint8_t> parse_hex_string(const std::string& hex_string)
{
    std::vector<uint8_t> result;
    std::istringstream in(hex_string);
    in >> std::hex;

    uint32_t value;
    while (in >> value)
    {
        assert(value <= 255U && "Invalid hex byte in string");
        result.push_back(static_cast<uint8_t>(value));
    }

    return result;
}

/// Compares two hex::dump objects to see whether they are
/// equal, i.e. they contain exactly the same data.
///
/// @param a The first hex::dump object
/// @param b The second hex::dump object
/// @return True if the hex::dump objects contain the same data,
///         otherwise false.
inline bool operator==(const dump& a, const dump& b)
{
    if (a.m_size != b.m_size)
    {
        return false;
    }

    // They have the same size - do they point to the same data?
    if (a.m_data == b.m_data)
    {
        return true;
    }

    // It is two different buffers - is the content equal?
    return std::equal(a.m_data, a.m_data + a.m_size, b.m_data);
}

/// The actual output operator which prints the data buffer to
/// the choosen output stream.
///
/// @param out The output stream
///
/// @param hex The hexdump struct initialized with the data and size
///        that should be printed
///
/// @return the used output stream
inline std::ostream& operator<<(std::ostream& out, const dump& hex)
{
    const uint8_t* data = hex.m_data;
    uint32_t size = std::min(hex.m_size, hex.m_max_size);
    std::string prefix = hex.m_prefix;
    std::string postfix = hex.m_postfix;

    assert(data);
    assert(size > 0);

    // don't change formatting for out
    std::ostream s(out.rdbuf());
    s << std::hex << std::setfill('0');

    std::string buf;
    buf.reserve(17); // premature optimization

    for (uint32_t i = 0; i < size; ++i)
    {
        if ((i % 16) == 0)
        {
            if (i != 0)
            {
                s << " " << buf << std::endl;
                buf.clear();
            }
            s << std::setw(4) << i << "  ";
        }

        uint8_t c = data[i];
        s << prefix << std::setw(2) << (uint32_t) c << postfix;
        // write the charater if it's writeable, otherwise write a "."
        buf += (0x20 <= c && c <= 0x7e) ? c : '.';
    }

    if (size % 16 != 0)
    {
        // If size if not a multiple of 16 there will be some empty
        // columns in our print out. The remainder i.e. 16 - (size % 1&)
        // e.g. for some basic cases:
        //
        // size = 15
        // remainder = 16 - (15 % 16) = 1
        //
        // size = 17
        // remainder = 16 - (17 % 16) = 15
        //
        // and so forth.
        uint32_t remainder = 16 - (size % 16);

        // We add 3 space for each missing character in the output
        auto element_size = (2 + prefix.length() + postfix.length());
        s << std::string(element_size * remainder, ' ');
    }

    s << " " << buf << std::endl;

    if (size < hex.m_max_size && 16 < hex.m_max_size)
    {
        uint32_t last_row = (hex.m_max_size / 16) * 16;

        s << "...." << std::endl;
        s << std::setw(4) << last_row << std::endl;
    }

    return out;
}
}
